core/src/main/java/io/grpc/transport/MessageDeframer.java

```diff
@@ -124,27 +148,131 @@ public MessageDeframer(Listener listener) {
     Preconditions.checkArgument(numMessages > 0, "numMessages must be > 0");
     if (isClosed()) {
       return;
     }
-    pendingDeliveries += numMessages;
-    deliver();
+    // Make the number of requested messages visible to the transport thread.
+    transportCount.addAndGet(numMessages);
+    requestedMessageCount += numMessages;
+    if (messagesAvailable && !appInRequest) {
+      appInRequest = true;
+      try {
+        listener.messagesAvailable(this);
+      } finally {
+        appInRequest = false;
+      }
+    }
+  }
+
+  /**
+   * Called in the application thread to consume the available messages provided by the transport.
+   * @param consumer which receives all available messages.
+   */
+  @Override
+  public int drainTo(StreamListener.MessageConsumer consumer) {
+    int count = 0;
+    do {
+      if (appQueue == null) {
+        appQueue = new ArrayDeque<Frame>();
+      }
+      if (appQueue.isEmpty()) {
+        // Get the queue from the transport and copy it into the appQueue
+        ArrayDeque<Frame> transportQueue;
+        do {
+          transportQueue = queueRef.getAndSet(null);
+          if (transportQueue != null) {
+            appQueue.addAll(transportQueue);
+            transportQueue.clear();
+          }
+          // Return it to the transport thread
+        } while (!queueRef.compareAndSet(null, transportQueue));
+        messagesAvailable = !appQueue.isEmpty();
+      } else {
+        // Consume the messages we already have. Will check for messagesAvailable
+        count += drainTo(consumer, appQueue);
+      }
+    } while (requestedMessageCount > 0 && messagesAvailable);
+    return count;
+  }
+
+  private int drainTo(StreamListener.MessageConsumer consumer, ArrayDeque<Frame> queue) {
+    int count = 0;
+    int consumedBytes = 0;
+    boolean endOfStream = false;
+    try {
+      Frame frame;
+      do {
+        frame = queue.peek();
+        if (frame == null) {
+          // Nothing to read
+          break;
+        }
+        if (frame == END_OF_STREAM) {
+          // Transport enqueued EOS, so empty queue and report to the listener
+          queue.clear();
+          endOfStream = true;
+          break;
+        }
+        // Can't deliver messages to the consumer or return flow control
+        if (requestedMessageCount == 0) {
+          break;
+        }
+        // Consume the message from the queue
+        queue.poll();
+        consumedBytes += frame.consumedBytes;
+        if (frame.stream != null) {
+          try {
+            // Calls to consumer may trigger a call to request in the same thread, we prevent
+            // the unnecessary call to notify again as this loop will pick up the additional
+            // count.
+            appInRequest = true;
+            count++;
+            consumer.accept(frame.stream);
+          } catch (Throwable t) {
+            t.printStackTrace();
+            // Decision point: Log / exit / drain the available messages
```

Comment:
If the exception was just passed along, how bad of a state would the Deframer be left in?  Also, since this is coming from client code, its probably okay to let the application thread die.  As your comment points out, this code is not in a good position to make a decision about the error.  

Ideally, the application would get notified of the failure, (exceptions should be expected and prepared for right?) and the Deframer, Call, and Channel would still be usable afterwards.  This may not be possible, and if not, the worst I can see happening is the Channel being shutdown.

Logging is too weak IMO, and likely not going to convey the severity of the problem.  You also risk logspamming the server if these can happen too quickly.